File: android/app/src/main/kotlin/com/narraleaf/shell/ShellConfig.kt
package com.narraleaf.shell

import android.content.pm.ActivityInfo
import android.content.res.AssetManager
import org.json.JSONObject

/**
 * The per-game settings the repacker injects as `assets/shell-config.json`.
 *
 * Parsed with the platform's own JSON reader — the shell carries no third-party
 * libraries. Every field falls back to a default, so an older template paired
 * with a newer writer (or the reverse) degrades instead of crashing.
 */
data class ShellConfig(
    val schemaVersion: Int,
    val orientation: Orientation,
    val backgroundColor: Int,
    /**
     * Opaque token the payload decoder needs, injected only when the packer
     * encoded this build's payload. Null means the payload is plain and is
     * served as-is — one shell runs both kinds of game.
     */
    val contentKey: String?,
) {

    enum class Orientation {
        LANDSCAPE,
        PORTRAIT,
        AUTO;

        /** The ActivityInfo constant this maps onto at runtime. */
        fun toActivityInfo(): Int = when (this) {
            LANDSCAPE -> ActivityInfo.SCREEN_ORIENTATION_SENSOR_LANDSCAPE
            PORTRAIT -> ActivityInfo.SCREEN_ORIENTATION_SENSOR_PORTRAIT
            AUTO -> ActivityInfo.SCREEN_ORIENTATION_UNSPECIFIED
        }

        companion object {
            fun parse(value: String?): Orientation = when (value?.lowercase()) {
                "portrait" -> PORTRAIT
                "auto" -> AUTO
                else -> LANDSCAPE
            }
        }
    }

    companion object {
        const val PATH = "shell-config.json"

        /** Matches the pre-boot background the web export paints. */
        private const val DEFAULT_BACKGROUND = 0xFF000000.toInt()

        val DEFAULT = ShellConfig(
            schemaVersion = 1,
            orientation = Orientation.LANDSCAPE,
            backgroundColor = DEFAULT_BACKGROUND,
            contentKey = null,
        )

        /**
         * Read the injected config, falling back to [DEFAULT] when it is
         * absent or unreadable — a template running without a repack (the CI
         * smoke test) still has to boot.
         */
        fun load(assets: AssetManager): ShellConfig {
            val text = runCatching {
                assets.open(PATH).bufferedReader().use { it.readText() }
            }.getOrNull() ?: return DEFAULT

            val json = runCatching { JSONObject(text) }.getOrNull() ?: return DEFAULT
            return ShellConfig(
                schemaVersion = json.optInt("schemaVersion", DEFAULT.schemaVersion),
                orientation = Orientation.parse(json.optString("orientation", null)),
                backgroundColor = parseColor(json.optString("backgroundColor", null))
                    ?: DEFAULT.backgroundColor,
                contentKey = json.optString("contentKey", null)?.takeIf { it.isNotEmpty() },
            )
        }

        /** #rgb / #rrggbb / #aarrggbb, or null when unparseable. */
        private fun parseColor(value: String?): Int? {
            val text = value?.trim()?.removePrefix("#") ?: return null
            val hex = when (text.length) {
                3 -> text.map { "$it$it" }.joinToString("")
                6, 8 -> text
                else -> return null
            }
            val parsed = hex.toLongOrNull(16) ?: return null
            return if (hex.length == 6) (0xFF000000L or parsed).toInt() else parsed.toInt()
        }
    }
}

File: ios/Shell/ShellConfig.swift
import UIKit

/// The per-game settings the repacker injects as `shell-config.json` beside the
/// payload.
///
/// Decoded with the platform's own JSON reader — the shell carries no
/// third-party libraries. Every field falls back to a default, so an older
/// template paired with a newer writer (or the reverse) degrades instead of
/// crashing.
struct ShellConfig {

    enum Orientation: String {
        case landscape
        case portrait
        case auto

        var mask: UIInterfaceOrientationMask {
            switch self {
            case .landscape: return .landscape
            case .portrait: return [.portrait, .portraitUpsideDown]
            case .auto: return .all
            }
        }

        static func parse(_ value: String?) -> Orientation {
            guard let value, let parsed = Orientation(rawValue: value.lowercased()) else {
                return .landscape
            }
            return parsed
        }
    }

    let schemaVersion: Int
    let orientation: Orientation
    let backgroundColor: UIColor
    /// Opaque token the payload decoder needs, injected only when the packer
    /// encoded this build's payload. Nil means the payload is plain and is
    /// served as-is — one shell runs both kinds of game.
    let contentKey: String?

    static let path = "shell-config.json"

    /// Matches the pre-boot background the web export paints.
    static let defaultConfig = ShellConfig(
        schemaVersion: 1,
        orientation: .landscape,
        backgroundColor: .black,
        contentKey: nil
    )

    /// Read the injected config, falling back to `defaultConfig` when it is
    /// absent or unreadable — a template running without a repack (the CI smoke
    /// test) still has to boot.
    static func load(bundle: Bundle = .main) -> ShellConfig {
        guard
            let url = bundle.url(forResource: "shell-config", withExtension: "json"),
            let data = try? Data(contentsOf: url),
            let object = try? JSONSerialization.jsonObject(with: data),
            let json = object as? [String: Any]
        else {
            return defaultConfig
        }
        let contentKey = (json["contentKey"] as? String).flatMap { $0.isEmpty ? nil : $0 }
        return ShellConfig(
            schemaVersion: json["schemaVersion"] as? Int ?? defaultConfig.schemaVersion,
            orientation: Orientation.parse(json["orientation"] as? String),
            backgroundColor: parseColor(json["backgroundColor"] as? String) ?? defaultConfig.backgroundColor,
            contentKey: contentKey
        )
    }

    /// #rgb / #rrggbb / #aarrggbb, or nil when unparseable.
    private static func parseColor(_ value: String?) -> UIColor? {
        guard var hex = value?.trimmingCharacters(in: .whitespacesAndNewlines) else { return nil }
        if hex.hasPrefix("#") { hex.removeFirst() }
        if hex.count == 3 { hex = hex.map { "\($0)\($0)" }.joined() }
        guard hex.count == 6 || hex.count == 8, let parsed = UInt32(hex, radix: 16) else { return nil }

        let hasAlpha = hex.count == 8
        let a = hasAlpha ? CGFloat((parsed >> 24) & 0xff) / 255 : 1
        let r = CGFloat((parsed >> 16) & 0xff) / 255
        let g = CGFloat((parsed >> 8) & 0xff) / 255
        let b = CGFloat(parsed & 0xff) / 255
        return UIColor(red: r, green: g, blue: b, alpha: a)
    }
}
